Allow plugins to use SLF4J for logging

stephan-gh opened the pull request:
This is a set of 3 commits to improve the logging implementation of Paper and allow plugins to use SLF4J as alternative logging API. The exact changes and the reasoning for them is explained in the commit messages (either directly or in the patches) so I'm not going to repeat everything here.

SLF4J is a commonly used abstraction for various logging frameworks and it is the easiest way to handle logging in an implementation-agnostic way. It is also used by many Java libraries as logging abstraction. I intend to create a similar PR for Waterfall so eventually, the same logging API can be used on Paper, Waterfall, Sponge and other modding APIs.

The primary difficulty to make this work was to move the plugin prefixes out of a custom JUL `PluginLogger` implementation and handle them directly in the underlying logging framework (Log4J). Otherwise, messages logged with SLF4J would be missing the plugin prefix.

The solution is pretty simple: Use the plugin name as logger name and make it visible in the log using the Log4J configuration. Hide the logger name for Minecraft and Paper log messages to make it less verbose.

stephan-gh committed:
Use Log4j IOStreams to redirect System.out/err to logger

Log4j2 provides an optimized implementation of PrintStream that
redirects its output to a logger. Use it instead of a custom
implementation for minor performance improvements and some fixes.

With the old implementation, each call to System.print()
results in a separate line, even though it should not result in
a line break. Log4j's implementation handles it correctly.

stephan-gh committed:
Handle plugin prefixes in Log4j configuration

stephan-gh committed:
Allow plugins to use SLF4J for logging

SLF4J is a commonly used abstraction for various logging frameworks
such as java.util.logging (JUL) or Log4j. Currently, plugins are
required to do all their logging using the provided JUL logger.
This is annoying for plugins that target multiple platforms or when
using libraries that log messages using SLF4J.

Expose SLF4J as optional logging API for plugins, so they can use
it without having to shade it in the plugin and going through
several layers of logging abstraction.

kashike commented on Spigot-API-Patches/0067-Handle-plugin-prefixes-in-implementation-logging-con.patch:
```diff
@@ -0,0 +1,43 @@
+     private FileConfiguration newConfig = null;
+     private File configFile = null;
+-    private PluginLogger logger = null;
++    private Logger logger = null; // Paper: PluginLogger -> Logger
```
`// Paper - PluginLogger -> Logger`

kashike commented on Spigot-API-Patches/0067-Handle-plugin-prefixes-in-implementation-logging-con.patch:
```diff
@@ -0,0 +1,43 @@
+         this.classLoader = classLoader;
+         this.configFile = new File(dataFolder, "config.yml");
+-        this.logger = new PluginLogger(this);
++        this.logger = Logger.getLogger(description.getName()); // Paper: Handle plugin prefix in implementation
```
`// Paper - Handle plugin prefix in implementation`

kashike replied:
Should be `getPrefix()`.

kashike commented on Spigot-Server-Patches/0238-Use-Log4j-IOStreams-to-redirect-System.out-err-to-lo.patch:
```diff
@@ -0,0 +1,51 @@
+             <scope>runtime</scope>
+         </dependency>
+ 
++        <!-- Paper: Add additional Log4J dependencies -->
```
`<!-- Paper - Add additional Log4J dependencies -->`

kashike commented on Spigot-Server-Patches/0239-Handle-plugin-prefixes-using-Log4J-configuration.patch:
```diff
@@ -0,0 +1,231 @@
+                                 </relocation>
+                             </relocations>
+                             <transformers>
++                                <!-- Paper: Exclude log4j2-test.xml -->
```
`<!-- Paper - Exclude log4j2-test.xml -->`

kashike commented on Spigot-API-Patches/0068-Allow-plugins-to-use-SLF4J-for-logging.patch:
```diff
@@ -0,0 +1,54 @@
+             <scope>compile</scope>
+         </dependency>
+ 
++        <!-- Paper: Add SLF4J -->
```
`<!-- Paper - Add SLF4J -->`

kashike commented on Spigot-Server-Patches/0239-Handle-plugin-prefixes-using-Log4J-configuration.patch:
```diff
@@ -0,0 +1,231 @@
+     {
+         playerSample = getInt( "settings.sample-count", 12 );
+-        System.out.println( "Server Ping Player Sample Count: " + playerSample );
++        Bukkit.getLogger().log( Level.INFO, "Server Ping Player Sample Count: {0}", playerSample ); // Paper: Use logger
```
`// Paper - Use logger`

stephan-gh committed:
Add workaround for plugins modifying the parent of the plugin logger

Essentials uses a custom logger name ("Essentials") instead of the
plugin logger. Log messages are redirected to the plugin logger by
setting the parent of the "Essentials" logger to the plugin logger.

With our changes, the plugin logger is now also called "Essentials",
resulting in an infinite loop. Make sure plugins can't change the
parent of the plugin logger to avoid this.

stephan-gh committed:
Disable logger prefix for various plugins bypassing the plugin logger

Some plugins bypass the plugin logger and add the plugin prefix
manually to the log message. Since they use other logger names
(e.g. qualified class names) these would now also appear in the
log. Disable the logger prefix for these plugins so the messages
show up correctly.
